src/frequenz/client/dispatch/test/service.py

```diff
@@ -0,0 +1,259 @@
+                        getattr(request.update, split_path[0])
+                    )
+                case "recurrence":
+                    match split_path[1]:
+                        case "end_criteria":
+                            pb_dispatch.recurrence.end_criteria.CopyFrom(
+                                request.update.recurrence.end_criteria
+                            )
+                        case "freq" | "interval":
+                            setattr(
+                                pb_dispatch.recurrence,
+                                split_path[1],
+                                getattr(request.update.recurrence, split_path[1]),
+                            )
+                        # Fields of type list that need to be copied
+                        case (
+                            "byminutes"
+                            | "byhours"
+                            | "byweekdays"
+                            | "bymonthdays"
+                            | "bymonths"
+                        ):
+                            getattr(pb_dispatch.recurrence, split_path[1])[:] = getattr(
+                                request.update.recurrence, split_path[1]
+                            )[:]
+
+        dispatch = Dispatch.from_protobuf(pb_dispatch)
+        dispatch.update_time = datetime.now(tz=timezone.utc)
+
+        return Empty()
+
+    async def GetMicrogridDispatch(
+        self,
+        request: DispatchGetRequest,
+    ) -> PBDispatch:
+        """Get a single dispatch."""
+        dispatch = next((d for d in self.dispatches if d.id == request.id), None)
+
+        if dispatch is None:
+            error = grpc.RpcError()
+            # pylint: disable=protected-access
+            error._code = grpc.StatusCode.NOT_FOUND  # type: ignore
+            error._details = "Dispatch not found"  # type: ignore
+            # pylint: enable=protected-access
+            raise error
+
+        return dispatch.to_protobuf()
+
+    async def DeleteMicrogridDispatch(
+        self,
+        request: DispatchDeleteRequest,
+    ) -> Empty:
+        """Delete a given dispatch."""
+        num_dispatches = len(self.dispatches)
+        self.dispatches = [d for d in self.dispatches if d.id != request.id]
+
+        if len(self.dispatches) == num_dispatches:
+            error = grpc.RpcError()
+            # pylint: disable=protected-access
+            error._code = grpc.StatusCode.NOT_FOUND  # type: ignore
+            error._details = "Dispatch not found"  # type: ignore
+            # pylint: enable=protected-access
+            raise error
+
+        return Empty()
+
+    # pylint: enable=invalid-name
+
+
+class MockClient(Client):
+    """Mock client for testing."""
+
+    def __init__(self) -> None:
+        """Initialize the mock client."""
+        super().__init__(grpc.aio.insecure_channel("mock"), "mock")
+        self._stub = Service()  # type: ignore
```

Comment:
So, if I understand this correctly, users to add fake dispatches to the mock service/client need to do something like:

```py
mock_client._stub.dispatches.append()
```
?

If so, I think you should provide some `add_dispatch()` and `remove_dispatch()` or so to the mock client, or a R/W `@property` to access the underlying `dispatches` list as `mock_client.dispatches` directly.

Reply:
No, it's way simpler. You can use the client directly, just the normal create/list/update methods.
All the user needs to do is create an instance of `mock_client` really and can use the methods directly.

Reply:
Oh, ok, I see. For me this is not a mock at all and I wouldn't call it that. A mock in general is super shallow, has almost no logic except for registering which stuff you used from it so you can then assert that what you expected to happen really happened. I think this should be called a fake or simulation. And I wouldn't encourage it's use for **unit tests**.

> Fakes are objects that have working implementations. On the other hand, mocks are objects that have predefined behavior.

https://medium.com/@mohammad.roshandelpoor/whats-the-difference-between-faking-mocking-and-stubbing-2f5aab410531

Seems really cool for pseudo-integration tests though (as is not a pure integration test either, as it's not the real server).

But this also makes me think why not using the real server for this instead.

Again, probably a bigger topic affecting all the APIs.

Reply:
Well, they are very useful for unit tests, it's easier to construct the data for the expected response this way instead of mocking up the exact response each time.. but I can change the code to do that instead. 

Note that this is exactly what we talked about already in slack, I specifically asked about exactly this ;)

Reply:
Haha, yeah, but sometimes you don't realize until you see the idea in the largest context. Also it is different when something is internal than when you make it public though.

So, for narrowing things down to the context of this PR, I would just rename `MockClient` to `FakeClient` and not expose the `Service` class at all, users of the clients don't need to know about protobuf stuff.

About:
> Well, they are very useful for unit tests, it's easier to construct the data for the expected response this way instead of mocking up the exact response each time.. but I can change the code to do that instead.

I understand it might be more convenient but I think is the wrong approach **for unit tests**. Why should a **unit test** for listing dispatches should fail if the create method of the client is broken?

Reply:
So you're concerned about the missing separation, that's fair. Though I do think I do access the internals of the FakeServer for that check in test_create ;)

Reply:
I didn't get that.

Reply:
BTW and FWIW, I don't think I'm being inconsistent with what I said on slack:

> I can't make an informed opinion, I feel I need getting my hands dirty first too to judge. My gut feeling says that it is probably wise to start with less abstraction (just mocking every function in every test) but at some point we'll find patterns and will want to abstract it away.
> But usually from the unit test point of view, I think when you mock in the test itself it is much more obvious what you are testing exactly, which is useful, so at some point is a matter of better readability and more repetition or the other way around.

Reply:
> I understand it might be more convenient but I think is the wrong approach for unit tests. Why should a unit test for listing dispatches should fail if the create method of the client is broken?

Alright, true, but this argument only counts for the tests of the client itself
Elegantely, I dodged the scenario you described by [combining](https://github.com/frequenz-floss/frequenz-client-dispatch-python/blob/v0.x.x/tests/test_dispatch_client.py#L383) those specific tests XD

Though the whole fake server is still useful for testing things that build on top of the client like the dispatch actor.